incuna_auth/middleware/permission.py

```diff
@@ -0,0 +1,232 @@
+
+        The default implementation returns [r'^'] - all URLs are protected.
+
+        The regexes will need to be compiled. (The utils.compile_urls method helps here.)
+        Override this method to supply your own list of protected URLs (for instance, from
+        Django settings) to this middleware.
+        """
+        return ALL_URLS
+
+    def is_resource_protected(self, request, **kwargs):
+        """
+        Returns true if and only if the resource's URL is *not* exempt and *is* protected.
+        """
+        exempt_urls = self.get_exempt_url_patterns()
+        protected_urls = self.get_protected_url_patterns()
+        path = request.path_info.lstrip('/')
+
+        path_is_exempt = any(m.match(path) for m in exempt_urls)
+        if path_is_exempt:
+            return False
+
+        path_is_protected = any(m.match(path) for m in protected_urls)
+        if path_is_protected:
+            return True
+
+        return False
+
+
+class FeinCMSPermissionMiddleware(BasePermissionMiddleware):
+    """
+    Middleware that allows or denies access based on the resource's access state.
+
+    Access state is provided by incuna_auth.models.AccessStateExtensionMixin and is used
+    as a marker to determine how a FeinCMS resource should be access-controlled.
+
+    The reasoning behind this is twofold. FeinCMS resources have unpredictable URLs, so
+    protecting them with the URL-based middleware is risky. Additionally, a client might
+    want to add a variety of permissions on their Pages (or similar) that don't
+    necessarily correspond to their location within the site. Adding an access_state
+    and checking that through middleware allows maximum flexibility and puts permissions
+    on the page level fully within the user's control.
+
+    This class protects all pages with an access_state of STATE_AUTH_ONLY. To protect
+    a different state or list of states, override get_protected_states.
+    """
+    def get_protected_states(self):
+        """
+        Returns a list of access states this middleware should apply to.
+
+        By default, returns STATE_AUTH_ONLY, which is the only non-custom access state
+        that implies any restriction.
+        """
+        return [AccessState.STATE_AUTH_ONLY]
+
+    def _get_resource_access_state(self, request):
+        """
+        Returns the FeinCMS resource's access_state, following any INHERITed values.
+
+        Will return None if the resource has an access state that should never be
+        protected. It should not be possible to protect a resource with an access_state
+        of STATE_ALL_ALLOWED, or an access_state of STATE_INHERIT and no parent.
+        """
+        feincms_page = getattr(request, 'feincms_page')
+
+        # Chase inherited values up the tree of inheritance.
+        INHERIT = AccessState.STATE_INHERIT
+        while (
+            feincms_page.access_state == INHERIT and
+            getattr(feincms_page, 'parent', None)
```

Comment:
This could be:
```python
        while feincms_page.access_state == INHERIT and hasattr(feincms_page, 'parent'):
            feincms_page = feincms_page.parent
```